models/repo.go

```diff
@@ -1605,92 +1618,79 @@ func DeleteRepository(doer *User, uid, repoID int64) error {
+		if _, err = e.Exec("UPDATE `repository` SET num_forks=num_forks-1 WHERE id=?", repo.ForkID); err != nil {
 			return fmt.Errorf("decrease fork count: %v", err)
 		}
 	}
 
-	if _, err = sess.Exec("UPDATE `user` SET num_repos=num_repos-1 WHERE id=?", uid); err != nil {
+	if _, err = e.Exec("UPDATE `user` SET num_repos=num_repos-1 WHERE id=?", uid); err != nil {
 		return err
 	}
 
 	if len(repo.Topics) > 0 {
-		if err = removeTopicsFromRepo(sess, repo.ID); err != nil {
+		if err = removeTopicsFromRepo(e, repo.ID); err != nil {
 			return err
 		}
 	}
 
-	projects, _, err := getProjects(sess, ProjectSearchOptions{
+	projects, _, err := getProjects(e, ProjectSearchOptions{
 		RepoID: repoID,
 	})
 	if err != nil {
 		return fmt.Errorf("get projects: %v", err)
 	}
 	for i := range projects {
-		if err := deleteProjectByID(sess, projects[i].ID); err != nil {
+		if err := deleteProjectByID(e, projects[i].ID); err != nil {
 			return fmt.Errorf("delete project [%d]: %v", projects[i].ID, err)
 		}
 	}
 
 	// FIXME: Remove repository files should be executed after transaction succeed.
 	repoPath := repo.RepoPath()
-	removeAllWithNotice(sess, "Delete repository files", repoPath)
+	removeAllWithNotice(e, "Delete repository files", repoPath)
 
-	err = repo.deleteWiki(sess)
+	err = repo.deleteWiki(e)
 	if err != nil {
 		return err
 	}
 
 	// Remove LFS objects
 	var lfsObjects []*LFSMetaObject
-	if err = sess.Where("repository_id=?", repoID).Find(&lfsObjects); err != nil {
+	if err = e.Where("repository_id=?", repoID).Find(&lfsObjects); err != nil {
 		return err
 	}
 
 	for _, v := range lfsObjects {
-		count, err := sess.Count(&LFSMetaObject{Oid: v.Oid})
+		count, err := e.Count(&LFSMetaObject{Oid: v.Oid})
 		if err != nil {
 			return err
 		}
 		if count > 1 {
 			continue
 		}
 
-		removeStorageWithNotice(sess, storage.LFS, "Delete orphaned LFS file", v.RelativePath())
+		removeStorageWithNotice(e, storage.LFS, "Delete orphaned LFS file", v.RelativePath())
 	}
 
-	if _, err := sess.Delete(&LFSMetaObject{RepositoryID: repoID}); err != nil {
+	if _, err := e.Delete(&LFSMetaObject{RepositoryID: repoID}); err != nil {
 		return err
 	}
 
 	if repo.NumForks > 0 {
-		if _, err = sess.Exec("UPDATE `repository` SET fork_id=0,is_fork=? WHERE fork_id=?", false, repo.ID); err != nil {
+		if _, err = e.Exec("UPDATE `repository` SET fork_id=0,is_fork=? WHERE fork_id=?", false, repo.ID); err != nil {
 			log.Error("reset 'fork_id' and 'is_fork': %v", err)
 		}
 	}
 
-	if err = sess.Commit(); err != nil {
-		sess.Close()
-		if len(deployKeys) > 0 {
-			// We need to rewrite the public keys because the commit failed
-			if err2 := RewriteAllPublicKeys(); err2 != nil {
-				return fmt.Errorf("Commit: %v SSH Keys: %v", err, err2)
-			}
-		}
-		return fmt.Errorf("Commit: %v", err)
-	}
-
-	sess.Close()
-
 	// We should always delete the files after the database transaction succeed. If
 	// we delete the file but the database rollback, the repository will be borken.
```

Comment:
Have you accounted for this?

Reply:
now this will be a bigger refactor ... I think I'll backport the check and use this pull to refactor it